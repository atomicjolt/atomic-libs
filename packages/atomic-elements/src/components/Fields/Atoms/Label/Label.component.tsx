import React from "react";
import styled from "styled-components";
import mixins from "../../../../styles/mixins";
import { ElementWrapperProps } from "../../../../types";
import { useContextProps } from "@hooks/useContextProps";
import { LabelContext } from "./Label.context";
import { useRenderProps } from "@hooks";

const StyledLabel = styled.label<{ $paddingBottom: string }>`
  ${mixins.Bold}
  display: block;
  font-size: 1.3rem;
  line-height: 1.1;
  color: var(--text-clr);

  ${({ $paddingBottom = "5px" }) =>
    $paddingBottom && `padding-bottom: ${$paddingBottom};`}
`;

export interface LabelProps
  extends ElementWrapperProps<React.LabelHTMLAttributes<HTMLLabelElement>> {
  as?: React.ElementType;
  $paddingBottom?: string;
}

/** A wrapper around an `<label />`  component with styles. */
export const Label = React.forwardRef(function Label(
  props: LabelProps,
  ref: React.Ref<HTMLLabelElement>
) {
  [props, ref] = useContextProps(LabelContext, props, ref);

  const { className, size, as = "label", ...rest } = props;

  const renderProps = useRenderProps({
    componentClassName: className,
    size,
  });

  return <StyledLabel as={as} ref={ref} {...renderProps} {...rest} />;
});
