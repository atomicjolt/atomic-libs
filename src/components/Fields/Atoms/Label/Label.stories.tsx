import { Meta, StoryObj } from "@storybook/react";
import { Label } from ".";

export default {
  title: "Fields/Atoms/Label",
  component: Label,
  parameters: {
    layout: "centered",
  },
} as Meta<typeof Label>;

type Story = StoryObj<typeof Label>;

export const Primary: Story = {
  args: {
    children: "Label",
  },
};
