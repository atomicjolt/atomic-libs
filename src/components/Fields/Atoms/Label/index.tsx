import styled from "styled-components";
import mixins from "../../../../styles/mixins";
import React from "react";
import classNames from "classnames";
import { ElementWrapperProps } from "../../../../types";

const StyledLabel = styled.label<{ $paddingBottom: string }>`
  ${mixins.Bold}
  display: block;
  font-size: 1.3rem;
  line-height: 1.1;
  color: var(--text-clr);

  ${({ $paddingBottom = "5px" }) =>
    $paddingBottom && `padding-bottom: ${$paddingBottom};`}
`;

export interface LabelProps
  extends ElementWrapperProps<React.LabelHTMLAttributes<HTMLLabelElement>> {
  as?: React.ElementType;
  $paddingBottom?: string;
}

/** A wrapper around an `<label />`  component with styles. */
export const Label = React.forwardRef(function Label(
  props: LabelProps,
  ref: React.Ref<HTMLLabelElement>
) {
  const { as, className, size = "auto", ...rest } = props;

  return (
    <StyledLabel
      as={as}
      ref={ref}
      className={classNames(className, `is-${size}`)}
      {...rest}
    />
  );
});
