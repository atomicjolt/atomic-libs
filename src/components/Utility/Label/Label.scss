@use "../../mixins.scss";

.aje-label {
  display: block;
  padding-bottom: 5px;
  @include mixins.bold;
  font-size: 1.3rem;
  line-height: 1.1;
  color: var(--text-clr);

  &--message,
  &--error {
    margin: 3px 0 0;
    @include mixins.bold;
    font-size: 1.3rem;
  }

  &--message {
    margin: 2px 0 0;
    @include mixins.regular;
    color: var(--text-clr-alt);
  }

  &--error {
    color: var(--error700);
  }
}
