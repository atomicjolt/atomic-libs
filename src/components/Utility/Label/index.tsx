import React from "react";
import cn from "classnames";
import { HasChildren } from "../../../types";

export interface Props {
  children: React.ReactNode;
  message?: string;
  htmlFor: string;
  hidden?: boolean;
  id?: string;
  error?: string;
}

export default function Label({
  children,
  message,
  htmlFor,
  hidden,
  id,
  error,
}: Props) {
  const classes = cn("aje-label", { "aje-hidden": hidden });

  return (
    <label className={classes} htmlFor={htmlFor} id={id}>
      {children}
      {message && <p className="aje-label--message">{message}</p>}
      {error && <ErrorLabel>{error}</ErrorLabel>}
    </label>
  );
}

interface ErrorLabelProps {
  children: React.ReactNode;
  id?: string;
}

export function ErrorLabel({ children, id }: ErrorLabelProps) {
  return (
    <p id={id} className="aje-label--error">
      {children}
    </p>
  );
}
