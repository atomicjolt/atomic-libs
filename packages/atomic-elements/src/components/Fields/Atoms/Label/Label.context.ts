import { createComponentContext } from "@utils/index";
import { LabelProps } from "./Label.component";

export const LabelContext = createComponentContext<LabelProps>();
