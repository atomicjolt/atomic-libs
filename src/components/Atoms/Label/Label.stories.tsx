import { Meta, StoryObj } from "@storybook/react";
import Label from ".";

const meta: Meta<typeof Label> = {
  title: "Atoms/Label",
  component: Label,
};

export default meta;

type Story = StoryObj<typeof Label>;

export const Primary: Story = {
  args: {
    children: "This is a Label",
    message: "This is a message beneath the label",
    error: "This is an error",
    hidden: false,
  },
};
