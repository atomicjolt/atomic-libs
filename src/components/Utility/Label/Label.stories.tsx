import { ComponentMeta, ComponentStory } from "@storybook/react";
import Label, { Props } from ".";

export default {
  title: "Utility/Label",
  component: Label,
} as ComponentMeta<typeof Label>;

const Template: ComponentStory<typeof Label> = (args: Props) => (
  <Label {...args} />
);

export const Default = Template.bind({});
Default.args = {
  children: "This is a Label",
  message: "This is a message beneath the label",
  error: "This is an error",
  hidden: false,
};

// Additional stories go here
