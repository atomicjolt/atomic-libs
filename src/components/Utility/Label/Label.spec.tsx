import React from "react";
import { render } from "@testing-library/react";
import Label from ".";

it("matches snapshot", () => {
  const result = render(<Label htmlFor="id">Label Content</Label>);
  expect(result.asFragment()).toMatchSnapshot();
});
