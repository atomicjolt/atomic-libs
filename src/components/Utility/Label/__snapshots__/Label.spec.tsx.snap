// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="aje-label"
    for="id"
  >
    Label Content
  </label>
</DocumentFragment>
`;
